src/plonk/lookup/prover.rs

```diff
@@ -94,82 +96,110 @@ impl<F: FieldExt> Argument<F> {
-                                .rotate_extended(&fixed_cosets[column_index], rotation)
+                    expression.evaluate_with(
+                        &mut evaluator_extended,
+                        &|_, scalar| poly::Ast::ConstantTerm(scalar),
+                        &|_, _| panic!("virtual selectors are removed during optimization"),
+                        &|evaluator, _, column_index, rotation| {
+                            evaluator
+                                .register_poly_as(
+                                    (Any::Fixed, column_index),
+                                    &fixed_cosets[column_index],
+                                )
+                                .with_rotation(rotation)
+                                .into()
                         },
-                        &|_, column_index, rotation| {
-                            pk.vk
-                                .domain
-                                .rotate_extended(&advice_cosets[column_index], rotation)
+                        &|evaluator, _, column_index, rotation| {
+                            evaluator
+                                .register_poly_as(
+                                    (Any::Advice, column_index),
+                                    &advice_cosets[column_index],
+                                )
+                                .with_rotation(rotation)
+                                .into()
                         },
-                        &|_, column_index, rotation| {
-                            pk.vk
-                                .domain
-                                .rotate_extended(&instance_cosets[column_index], rotation)
+                        &|evaluator, _, column_index, rotation| {
+                            evaluator
+                                .register_poly_as(
+                                    (Any::Instance, column_index),
+                                    &instance_cosets[column_index],
+                                )
+                                .with_rotation(rotation)
+                                .into()
                         },
-                        &|a| -a,
-                        &|a, b| a + &b,
-                        &|a, b| a * &b,
-                        &|a, scalar| a * scalar,
+                        &|_, a| -a,
+                        &|_, a, b| a + b,
+                        &|_, a, b| a * b,
+                        &|_, a, scalar| a * scalar,
                     )
                 })
                 .collect();
 
             // Compressed version of expressions
-            let compressed_expression = unpermuted_expressions
-                .iter()
-                .fold(domain.empty_lagrange(), |acc, expression| {
-                    acc * *theta + expression
-                });
+            let compressed_expression = unpermuted_expressions.iter().fold(
+                poly::Ast::ConstantTerm(C::Scalar::zero()),
```

Comment:
Consider using the `.reduce(...).unwrap_or(zero)` pattern here, if it would help.

Reply:
Here it does not: we are using `.iter()` so we get an iterator of references, which means that `reduce()` attempts to use `&Ast` as the accumulator which doesn't work. We'd need to `.iter().cloned()` which would deep-clone all of the ASTs, whereas `fold()` adds a zero constant to the operation.